Bluetooth: Audio: Add broadcast source metadata update function

Thalley opened the pull request:
Add API to do the BAP Broadcast Source Metadata update procedure,
which updates the metadata of a broadcast source while
it is streaming.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

depends on https://github.com/zephyrproject-rtos/zephyr/pull/49946

Thalley committed:
Bluetooth: Audio: Add broadcast source metadata update function

Add API to do the BAP Broadcast Source Metadata update procedure,
which updates the metadata of a broadcast source while
it is streaming.

This is also makes all checks for the "head stream" the same.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Move common broadcast sink test code to function

Move the functionality that is common between the two
broadcast sink test to a common function.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/46204

Thalley commented:
Rebased to solve merge conflicts

asbjornsabo commented on subsys/bluetooth/audio/broadcast_source.c:
```diff
@@ -641,6 +680,117 @@ int bt_audio_broadcast_source_reconfig(struct bt_audio_broadcast_source *source,
+			      source->streams[i], source);
+
+		if (source->streams[i]->codec != codec) {
+			BT_ERR("streams[%zu] does have correct codec: %p (expected %p)",
```
"does _not_" ?

asbjornsabo commented on subsys/bluetooth/audio/broadcast_source.c:
```diff
@@ -641,6 +680,117 @@ int bt_audio_broadcast_source_reconfig(struct bt_audio_broadcast_source *source,
+		return -EBADMSG;
+	}
+
+	/* All streams share the same codec, so we can just one of the streams */
```
A verb missing, I think.  "So we can just _use_ one of the streams"?

asbjornsabo commented on subsys/bluetooth/audio/endpoint.h:
```diff
@@ -73,6 +73,7 @@ struct bt_audio_unicast_group {
 struct bt_audio_broadcast_source {
 	uint8_t stream_count;
 	uint8_t subgroup_count;
+	bool adv_started;
```
Where is this one used?
In the functions above, a local variable seems to be used.

Thalley replied:
Ah, good catch. I started implementing with putting it here, but then figured that it better to extract the state from the endpoint states.

Thalley committed:
Bluetooth: Audio: Remove const for bt_audio_stream_reconfig codec

The codec is assigned to the stream. However since the metadata,
which is stored in the codec, may be modified by other means,
the stored pointer cannot be const, and thus the
codec argument to bt_audio_stream_reconfig cannot be const
either.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
This PR now depends on https://github.com/zephyrproject-rtos/zephyr/pull/49946, and thus converted to a draft

Thalley commented:
Rebased on https://github.com/zephyrproject-rtos/zephyr/pull/51897

asbjornsabo reviewed (approved):
LGTM

MariuszSkamra commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/broadcast_sink_test.c:
```diff
@@ -235,6 +249,16 @@ static void test_main(void)
+	/* Ensure that we also see the metadata update */
+	printk("Waiting for metadata update\n");
+	WAIT_FOR_FLAG(flag_base_metadata_updated)
+
```
redundant new line
